extend conan info command to show path information and filter by packages

claasd opened the pull request:
A new path_info command was added to conan, taking a user and optional
reference, settings and options.
Returns the export, source, build and package folders for the given
configuration.

memsharded commented:
Thanks for your contribution, I think the goal can be legit, and I am not opposed to having this command. However I have some concerns regarding the consistency of the approach:

- You are getting the dependency graph starting from the reference you can check. This might not be the same as the one you are actually expanding in your dependency graph of your project, even if providing the same input settings and options.
- Intermediate packages in the graph might provide default options for its dependencies, changing those defaults.

So I think the necessary approach should do the same as the ``conan info`` command, and most likely they should share code. Maybe the paths could be some other fields to be filtered (like url, license, etc) and we could also filter by packages. Something like:
```bash
$ conan info package_path -r Pkg/0.1@user/channel
Pkg/0.1@user/channel:
     package_path: /some/path/here
$ conan info paths -r Pkg/0.1@user/channel
Pkg/0.1@user/channel:
     source_path: /some/path/here
     build_path: /some/path/here
     package_path: /some/path/here
```

What do yo think? Thanks again!

claasd commented:
Let me tell you wat I want to achieve. I would like to run C++ Code coverage using conan install (--build = myPkg) -> import -> test. I use bullseye as coverage tool. Thus, I need to tell bullseye beforehand where the build folder will be.

I thought about adding it to the info command; but the info command currently does not take a user, and thus is not able to get the build folder if a conanfile is given as command.

I will propose something based on  the info command and your comments.

memsharded commented:
Hi Claas,

Still not sure about why do you want to directly access the build folder, and your workflow (I haven't used bullseye). The ``info`` command should take the same args as install, I don't know what you mean about not taking an user.
Probably makes sense more an opposite approach, let conan report on builds, so that info can be used for other tools? Right now there is some ongoing work for something similar, integrating the Artifactory build-info. It is done with logging (conan install logs all the actions), and then extracting info (automatically) from the logs. Please tell if want to chat with us (slack?). Thanks!

lasote commented on conans/test/command/package_id_test.py:
```diff
@@ -0,0 +1,72 @@
+
+    def test_short_pathes(self):
+        if platform.system() == "Windows":
+            currentEnv = os.environ.get("CONAN_USER_HOME_SHORT", None)
```
Use `tools.environment_append({"CONAN_USER_HOME_SHORT": short_folder})` context manager to avoid finally

claasd commented:
For c++, the code must be annotated during build to allow for code coverage (at least bullseye does it this way). Thus, I need to tell bullseye wich code he should annotate (my UUT) and wich not. I need to point bullseye to the build folder of my UUT component. 
Currently, I need to do the following:
* prepare bullseye
* check out from scm
* conan install (dependencies) and import stuff
* conan build (in the checked out folder)
* copy some stuff around (to emulate packaging, very error prone)
* run the tests (needs copied stuff and imported stuff)

I already have a workflow for installing my UUT and testing it:

* create conanfile.txt with requires and imports
* call conan install
* run tests

 It would be straightforward to re-use it and tell bullseye beforehand where the sources are that should be annotated:

* create conanfile.txt with requires and imports
* call conan path_info
* prepare bullseye
* call conan install --build my_uut
* run tests

Additionally, the scm checkout may contain some stuff that is not in the package and should not be taken into account when calculating code coverage.

I currently have a local solution that adds the fields to the info command, I will update this PR so you can have a look at it.

memsharded commented:
Now I think I get it. The trick is that bullseye is intercepting cl and linker, then you need to tell in advance the build folder, so its scan next build when it is done.

It would be pretty cool, if the tool had a command line, you could call programmatically, then enabling coverage in your package builds could be easily done with scopes:

```def build(self):
      if self.scopes.coverage:
            self.run("bullseye --folder %s" % os.getcwd())
      # normal build here
```
Then you will not have to worry, just call ``conan install --build my_uut --scope my_uut:coverage=True``

But otherwise, yes, I understand the flow, and this command totally makes sense for that use case, so looking forward your PR update.

claasd commented:
Hi,

i updated this, now it is part of the info command.
The info command shows four new fields:
* exportFolder
* sourceFolder
* buildFolder
* packageFolder

this can be selected using the `--only`

Additionally, i added a `--package_filter` option, that only prints information about packages that match the filter.

for my specific use case, I can now call
```bash
conan info --only build_folder --package_filter MyUUT/version*
```
and search the response for buildFolder

claasd commented:
Hi, regarding bullseye and command line:
Using the scopes to call the bullseye tool is straightforward, but then I would be mixing package information and the use of a very specific external tool. If someone wants to use a different tools, that would have to be supported as well ...

I think it would be bad architecture regarding separation of concerns if I would call bullseye from inside the package receipt.

memsharded commented:
Regarding the command-line: good, it is valid reasoning, thanks for the feedback. The important point for me is that it can be done easily, so if someone is using bullseye in their flow, they can already integrate it in their recipes if they want. Sure, if other tools are to be supported, then they should be added to the recipe (maybe using a different scope for each one, you can have just a ``bullseye`` scope if you want). 

I like the decoupled approach for tools, but I also understand someone else considering the coverage tool just another build tool, exactly the same as the compiler or the build system, so they would be willing to fully integrate it in their recipes.

claasd commented:
There is one issue that crossed my mind: with `short_paths = True`, this will create a lot of folders, even if they are not used.

memsharded commented:
Do you want me to have a look to the ``short_paths`` issue, or are you already working on something? Why would those folders be created? Thanks!

claasd commented:
When getting the pathes, I use the methods from `SimplePath`. These are calling `_shortener`
This creates the pathes when `short_paths` is true, and alo creates the `.conan_link` file:
```python
  ...
    try:
        os.makedirs(short_home)
    except:
        pass
    redirect = tempfile.mkdtemp(dir=short_home, prefix="")
    # This "1" is the way to have a non-existing directory, so commands like
    # shutil.copytree() to it, works. It can be removed without compromising the
    # temp folder generator and conan-links consistency
    redirect = os.path.join(redirect, "1")
    save(link, redirect)
```
This, when calling info, the pathes will be created, even if they will never be used.

memsharded commented:
Yes, it seems an issue. I will have a look when possible and tell you about possible solutions. Thanks!
